Don't require nonce in JsonTransactionsStorage constructor

Krigpl opened the pull request.

Krigpl commented:
@cameel @kbeker 
What do you think?

cameel commented:
The solution with an `init()` method seems fine. I don't like the fact that an object for some time is hanging in an uninitialized state but I don't really have any better suggestion. Maybe adding some `assert`s to make sure that other methods never run before the nonce has been initialized would be a good idea.

I have some questions about the initialization in case where the nonce already exists:
- If the new nonce is greater than the existing one, an error is raised. What if there's an old JSON file on disk and then immediately the class gets initialized with a new, higher nonce from the blockchain? Shouldn't the higher nonce just be used instead?
- If the nonce is smaller than the one in the file, it's simply ignored. Won't this cause errors when we try to publish transactions using this nonce? Shouldn't an error be raised in this case?

Krigpl commented:
> The solution with an init() method seems fine. I don't like the fact that an object for some time is hanging in an uninitialized state but I don't really have any better suggestion. Maybe adding some asserts to make sure that other methods never run before the nonce has been initialized would be a good idea.

Yeah, I had the same concerns but also didn't come up with anything better.

Regarding existing non zero nonce at initialization - ideally the nonce would be zero first time we initialize the module with a new Ethereum address. But we had to make it compatible with existing Golem users thus those cases.
- It's fine if the nonce from the network is lower than the one in the storage. This can happen in various cases, like not yet mined transaction being dropped from the transactions pool or when you reconnect to a node that is not yet fully synced. In this case we just propagate existing transactions and use the nonce from the storage for new ones - we won't create duplicated nonce.
- Here is a pretty strong and fundamental assumption that this particular Ethereum address isn't being used outside the SCI module. Meaning that user cannot create any transactions by themselves (in case of Golem, outside Golem application - this would break Golem's integrity). The same assumption should hold true for Concent address. Given this assumption we know that we're the only entity that creates and sends transactions to the network, thus if the network nonce is higher than the one we expect it means the assumption is broken. In this case we can't really be sure how many transactions have been sent, etc, so we panic and refuse to work.

I'll move the `init` method implementation to the base class,

cameel commented:
OK. Thanks. That explains it. Maybe it would be a good idea to put this explanation in a comment in the code? I think it's not obvious.

> Here is a pretty strong and fundamental assumption that this particular Ethereum address isn't being used outside the SCI module. Meaning that user cannot create any transactions by themselves (in case of Golem, outside Golem application - this would break Golem's integrity). The same assumption should hold true for Concent address.

I can see a problem with this assumption in development. For production it's fine but right now we have a single Concent contract shared between three test clusters: dev, staging and testned. Since it's testnet it's not such a big deal but wouldn't it be better to have a separate instance of the contract for each environment? We asked about the possibility of deploying separate instances of the contract for each cluster but @lukasz-glen said there was some problem with that. Having a single one does violate that assumption though.

Krigpl commented:
> I can see a problem with this assumption in development. For production it's fine but right now we have a single Concent contract shared between three test clusters: dev, staging and testned. Since it's testnet it's not such a big deal but wouldn't it be better to have a separate instance of the contract for each environment? We asked about the possibility of deploying separate instances of the contract for each cluster but @lukasz-glen said there was some problem with that. Having a single one does violate that assumption though.

Yes, that's something I talked to @shadeofblue recently, I was proposing exactly this approach with separate contract instances for development. I don't know what @lukasz-glen meant though, I'll check that with him.

cameel commented:
OK. Thanks.

Krigpl commented:
@cameel 
I've moved the nonce checks into the base class and incorporated the other changes related to merging two functions into one. Let me know if you feel like any other modifications would make it easier for you.

cameel commented:
I can't think of any. It looks fine.

@kbeker @rwrzesien Please take a look.

cameel commented:
Oh, maybe one - raising custom exceptions rather than just `Exception` would be nice. But that's probably out of scope of this pull request.

kbeker commented:
I see that we will have to do some changes in our's storage but I think it won't be too hard :) Much better to have nonce initialization inside just like you already did. :+1:

Krigpl commented:
@kbeker v1.5.0 released.
